cranelift-codegen/meta/src/isa/x86/instructions.rs

```diff
@@ -307,5 +307,74 @@ pub fn define(
         .operands_out(vec![a]),
     );
 
+    let x = &operand("x", TxN);
+    let y = &operand_doc("y", &TxN.lane_of(), "New lane value");
+    let a = &operand("a", TxN);
+
+    ig.push(
+        Inst::new(
+            "x86_pinsr",
+            r#"
+        Insert ``y`` into ``x`` at lane ``Idx``.
+        The lane index, ``Idx``, is an immediate value, not an SSA value. It
+        must indicate a valid lane index for the type of ``x``.
+        "#,
+        )
+        .operands_in(vec![x, Idx, y])
+        .operands_out(vec![a]),
+    );
+
+    let FxN = &TypeVar::new(
+        "FxN",
+        "A SIMD vector type containing floats",
+        TypeSetBuilder::new()
+            .floats(Interval::All)
+            .simd_lanes(Interval::All)
+            .includes_scalars(false)
+            .build(),
+    );
+    let x = &operand("x", FxN);
+    let y = &operand_doc("y", &FxN.lane_of(), "New lane value");
+    let a = &operand("a", FxN);
+
+    ig.push(
+        Inst::new(
+            "x86_insertps",
+            r#"
+        Insert a lane of ``y`` into ``x`` at using ``Idx`` to encode both which lane the value is 
+        extracted from  and which it is inserted to. This is similar to x86_pinsr but inserts 
+        floats, which are already stored in an XMM register.
+        "#,
+        )
+        .operands_in(vec![x, Idx, y])
+        .operands_out(vec![a]),
+    );
+
+    let x = &operand("x", FxN);
+    let y = &operand("y", FxN);
+    let a = &operand("a", FxN);
+
+    ig.push(
+        Inst::new(
+            "x86_movsd",
+            r#"
+        Move the low 64 bits of the float vector ``y`` to the low 64 bits of float vector ``x``
```

Comment:
Isn't this equivalent to a sequence of copy + raw_bitcast from SIMD vector to scalar? If so, for the sake of keeping the instruction set maximally minimal, could we not introduce this new instruction?

Reply:
I think we could do this but I'm worried that the semantics might become confusing:

 - `copy` for f32/f64 is encoded as `MOVAPS` which moves all 128 bits from one register to another
 - `x86_movsd` is encoded as `MOVSD` which only moves the bottom 64 bits from one register to another, leaving the higher bits unchanged as long as the source is a register and not a memory location; this effectively inserts an f64 into lane 0 of a vector, which is how I'm using it to legalize `insertlane`

We could replace the encoding of `copy.f64` to use `MOVSD` (and `copy.f32` could use `MOVSS`) instead of `MOVAPS` but this leaves the upper bits unchanged, something which is not clearly reflected in the `copy` instruction documentation. This theoretically is fine as long as we maintain the assumption that the type system will protect us from uncleared high bits. One additional consideration: `MOVSD/MOVSS` are 3-byte opcodes whereas `MOVAPS` is a 2-byte opcode. Thoughts?

Reply:
> but this leaves the upper bits unchanged, something which is not clearly reflected in the copy instruction documentation. This theoretically is fine as long as we maintain the assumption that the type system will protect us from uncleared high bits.

For 8 and 16 bit values the higher bits can be anything too, because `ireduce` is a no-op.

Reply:
Sounds like @bjorn3 is ok with changing `copy.f32/f64` to `MOVSD/MOVSS`; @bnjbvr?

Reply:
Shoot... I implemented this change and then realized that cranelift's `copy` does not accept a destination operand so I can't control which XMM register to move the lower bits to. I'm leaving `x86_movsd` in for now.

I had another thought that might resolve the underlying issue of adding all these new x86-specific instructions. If I can get https://github.com/abrown/cranelift/pull/3 working, we then have the ability to express the following without any new instructions:

```
e.enc...(insertlane.bind(F64).bind(lane0), [recipe and opcodes for MOVSD]);
e.enc...(insertlane.bind(F64).bind(lane1), [recipe and opcodes for MOVLHPS]);
```

I guess that's more motivation to get that PR working. Should we leave `x86_movsd` in for now and refactor later?

Reply:
Sounds good to me, yeah. Thanks for trying!